### PR TITLE
Refactor moe.py PR #1

### PR DESCRIPTION
**Pull request by Shuwen-Fang**
# Description

Refactor sparse_matmul for implementing chunking feature to overlap AG-RS. The refactor extracts individual components (routing, gmm-up, gmm-down) from the monolithic wrapper function so it's more extensible and makes it possible to implement features like chunking. 

High level psuedocode for what the end state should look like: 

```
def moe_route_and_compute(x, ...):
    routing, route_metadata = route(logits)
    if not chunking:
        x = dispatch(x, route_metadata)
        intermediate = gmm_up(x, w0, w1, routing)
    else:
        intermediate = zeros(...)
        for x_chunk in chunks(x):                      
            x_chunk = dispatch(x_chunk, route_metadata)
            intermediate += gmm_up(x_chunk, w0_chunk, w1_chunk, routing)
    intermediate = maybe_tp_reduce(intermediate)
    output = gmm_down(intermediate, wo, routing)
    return unpermute(output, routing)
```

This PR refactors routing and gmm_up into helper functions, and creates Dataclass RouteOutput and RouteMetadata to hold the 11 return values.

Next steps:
- Extract gmm_down and unpermute into helpers; remote wrapper
- Implement the chunking loop (may required splitting route further)

# Tests

Testing
* Runing maxtext training succeeds:
- Sparse matmul:
```
source /home/shuwenf_google_com/venv-maxtext/bin/activate && PYTHONPATH=src python -m maxtext.trainers.pre_train.train \
  src/maxtext/configs/base.yml \
  model_name=deepseek3-tiny \
  base_output_directory=/tmp/moe_refactor_test \
  run_name=smoke_test \
  dataset_type=synthetic \
  enable_checkpointing=false \
  sparse_matmul=True \
  steps=3 \
  per_device_batch_size=1 \
  max_target_length=128 \
  ici_expert_parallelism=8 \
  attention=dot_product \
  2>&1 | tail -20
```

- dense matmul:
```
source /home/shuwenf_google_com/venv-maxtext/bin/activate && PYTHONPATH=/home/shuwenf_google_com/maxtext python -m maxtext.trainers.pre_train.train \
  src/maxtext/configs/base.yml \
  model_name=deepseek3-tiny \
  base_output_directory=/tmp/moe_refactor_test \
  run_name=dense_ep8 \
  dataset_type=synthetic \
  enable_checkpointing=false \
  sparse_matmul=False \
  steps=3 \
  per_device_batch_size=1 \
  max_target_length=128 \
  ici_expert_parallelism=8 \
  attention=dot_product \
  2>&1 | grep -E "completed step|Error|Traceback" | head -10
```

* Correctness & performance did not change, verified by running 
```
source /home/shuwenf_google_com/venv-maxtext/bin/activate && PYTHONPATH=src python -m maxtext.trainers.pre_train.train \
  src/maxtext/configs/base.yml \
  model_name=deepseek3-tiny \
  base_output_directory=/tmp/moe_refactor_test \
  run_name=smoke_test \
  dataset_type=synthetic \
  enable_checkpointing=false \
  sparse_matmul=True \
  steps=3 \
  per_device_batch_size=1 \
  max_target_length=128 \
  ici_expert_parallelism=8 \
  attention=dot_product \
  2>&1 | tail -20
```
on local branch vs main, found identical loss and negligible performance diff:
(main)
```
I0602 22:37:25.576668 131201031527552 metric_logger.py:207] completed step: 0, seconds: 7.605, TFLOP/s/device: 0.002, Tokens/s/device: 16.830, total_weights: 1024, loss: 12.268, lm_loss: 12.268, perplexity: 212713.656, moe_lb_loss: 0.000
I0602 22:37:25.578214 131201031527552 metric_logger.py:314] To see full metrics 'tensorboard --logdir=/tmp/moe_refactor_test/smoke_test/tensorboard/'
I0602 22:37:25.641861 131201031527552 metric_logger.py:207] completed step: 1, seconds: 0.141, TFLOP/s/device: 0.135, Tokens/s/device: 909.453, total_weights: 1024, loss: 12.174, lm_loss: 12.174, perplexity: 193703.422, moe_lb_loss: 0.000
I0602 22:37:25.709005 131201031527552 metric_logger.py:207] completed step: 2, seconds: 0.008, TFLOP/s/device: 2.310, Tokens/s/device: 15573.671, total_weights: 1024, loss: 12.124, lm_loss: 12.124, perplexity: 184197.062, moe_lb_loss: 0.000
```
(PR)
```
metric_logger.py:207] completed step: 1, seconds: 0.143, TFLOP/s/device: 0.133, Tokens/s/device: 898.126, total_weights: 1024, loss: 12.174, lm_loss: 12.174, perplexity: 193703.422, moe_lb_loss: 0.000
I0602 22:36:16.367546 126581215829120 metric_logger.py:207] completed step: 2, seconds: 0.009, TFLOP/s/device: 2.162, Tokens/s/device: 14581.909, total_weights: 1024, loss: 12.124, lm_loss: 12.124, perplexity: 184197.062, moe_lb_loss: 0.000
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/moe.py [hunk 1]
```diff
@@ -863,7 +886,7 @@ def local_permute(
         across all the batch shards. Shape `[num_batch_shards, num_experts].
       local_expert_size: The number of experts handled by the current shard.
       shard_index: The index of the current expert shard (0 to
-        num_expert_parallelism - 1).
+        self.get_expert_parallelism_size() - 1).
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py]: you don't need to update comments?

File: src/maxtext/layers/moe.py [hunk 2]
```diff
@@ -1455,42 +1462,69 @@ def get_active_sharding_axes(pspec_dim_axes, tensor_dim_index):
...
+            P(),  # Replicate the input key
+        ),
+        out_specs=(
+            self._logical_to_mesh_axes((batch_logical_axis, "activation_norm_length", "activation_embed")),
+            P(),  # Handle None or replicate the output
+            P(),  # Handle None or replicate the output
+        ),
+        check_vma=False,
+    )
+    def wrapper(x, logits, pre_bias_logits, w0, w1, wo, w0_bias, w1_bias, wo_bias, rngs):
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py:1537]: We should give this function a better name, like `sparse_matmul_core`?

File: src/maxtext/layers/moe.py [hunk 3]
```diff
@@ -1455,42 +1462,69 @@ def get_active_sharding_axes(pspec_dim_axes, tensor_dim_index):
...
+      batch_size, sequence_length, _ = x.shape
+      x, routing, route_metadata = route(x, logits, pre_bias_logits, rngs)
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py:1539]: It would be my dream if we can break the shardmap of wrapper into smaller shardmaps, e.g. separate shardmap on route function, gmm_up, and permute/unpermute! But not in this PR of course

File: src/maxtext/layers/moe.py [hunk 4]
```diff
@@ -1299,37 +1322,15 @@ def get_routed_moe_shardings(is_batch_sharded_by_expert):
...
-      if num_expert_parallelism > 1:
-        expert_shard_id = jax.lax.axis_index(self._expert_parallelism_name)
-      else:
-        expert_shard_id = 0
-      num_expert_parallelism = self.get_expert_parallelism_size()
-      global_group_sizes = None
+    def route(x, logits, pre_bias_logits, rngs):
+      """Route tokens by expert"""
+      num_ep = self.get_expert_parallelism_size()
+      expert_shard_id = jax.lax.axis_index(self._expert_parallelism_name) if num_ep > 1 else 0
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py:1340]: maybe else None? I imagine expert_shard_id won't ever get used when num_ep == 1.

File: src/maxtext/layers/moe.py (hunk 1 above)
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py]: Done

File: src/maxtext/layers/moe.py (hunk 2 above)
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py:1537]: thoughts on `moe_route_and_compute`? (prev suggestion by matt)

File: src/maxtext/layers/moe.py (hunk 3 above)
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py:1539]: added follow up bug

File: src/maxtext/layers/moe.py (hunk 4 above)
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py:1340]: roll_to_expert_id=num_experts_per_shard * expert_shard_id is called even when num_ep == 1 in the ring of experts = true path. 

Technically users should not configure ring of experts when ep = 1, but setting 0 would prevent a crash if that was set

File: src/maxtext/layers/moe.py [hunk 5]
```diff
@@ -1341,7 +1342,7 @@ def wrapper(x, logits, pre_bias_logits, w0, w1, wo, w0_bias, w1_bias, wo_bias, r
         )
 
         # "Route" tokens within each shard.
-        num_experts_per_shard = self.config.num_experts // num_expert_parallelism
+        num_experts_per_shard = self.config.num_experts // num_ep
         x, sorted_selected_experts, weights, group_sizes, selected_experts, lb_loss, bias_updates = self.permute(
```
[the remark this answers is not in the document]
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py:1358]: I don't believe there is a shape mismatch since all tests are passing

File: src/maxtext/layers/moe.py [hunk 6]
```diff
@@ -55,6 +56,28 @@
 COMBINE = "combine"
 
 
+@dataclasses.dataclass
```
[the remark this answers is not in the document]
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py]: updated

File: src/maxtext/layers/moe.py [hunk 7]
```diff
@@ -55,6 +56,28 @@
...
+
+  expert_shard_id: int
+  local_sorted_indices: Optional[jax.Array]  # batch-sharded EP only
+  all_shards_group_sizes: Optional[jax.Array]  # batch-sharded EP only
+  reshaped_group_sizes: Optional[jax.Array]  # non-batch-sharded EP only
+
+
+@dataclasses.dataclass
+class RouteOutput:
+  """Embed-independent routing state returned by route(). x and RouteMetadata are returned separately."""
```
[the remark this answers is not in the document]
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py]: updated

File: src/maxtext/layers/moe.py [hunk 8]
```diff
@@ -1547,22 +1581,23 @@ def get_active_sharding_axes(pspec_dim_axes, tensor_dim_index):
               # Mirror the ragged-prefix gather used in `local_permute`. The
               # un-permute can use the same valid-prefix length because the
               # routed token count is identical for forward and backward.
-              valid_end = jnp.sum(group_sizes).astype(jnp.int32)  # pylint: disable=undefined-variable
+              valid_end = jnp.sum(routing.group_sizes).astype(jnp.int32)  # pylint: disable=undefined-variable
```
[the remark this answers is not in the document]
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py]: updated

File: src/maxtext/layers/moe.py [hunk 9]
```diff
@@ -55,6 +57,28 @@
 COMBINE = "combine"
 
 
+@struct.dataclass
+class RouteMetadata:
+  """EP communication state needed to undo the forward all-to-all after expert computation."""
+
+  expert_shard_id: int
```
Comment (RissyRan) [inline src/maxtext/layers/moe.py:65]: Could you help add some comments for attributes?

File: src/maxtext/layers/moe.py [hunk 10]
```diff
@@ -55,6 +57,28 @@
...
+  local_sorted_indices: Optional[jax.Array]  # batch-sharded EP only
+  all_shards_group_sizes: Optional[jax.Array]  # batch-sharded EP only
+  reshaped_group_sizes: Optional[jax.Array]  # non-batch-sharded EP only
+
+
+@struct.dataclass
+class RouteOutput:
+  """Embed-independent routing state returned by route(). x and RouteMetadata are returned separately."""
+
+  group_sizes: jax.Array  # tokens per expert → gmm_fn
```
Comment (RissyRan) [inline src/maxtext/layers/moe.py]: Shall we add some comments for them?  i.e. `lb_loss: Optional[jax.Array]: # auxiliary loss for token distribution among experts`


FYI: https://github.com/AI-Hypercomputer/maxtext/blob/fb2a3f679604388d8ca4e9490bfab9f2d0d7c304/docs/reference/core_concepts/moe_configuration.md#L4

**Review by RissyRan — COMMENTED** [review]
Thanks for the change! Could you help run a few perf sanity checks to ensure no perf hit after the refactor? I think test for 1) FSDP & 2) FSDP + EP cases should be sufficient. You could test deepseek v2 locally between with/without changes.

File: src/maxtext/layers/moe.py (hunk 2 above)
Comment (RissyRan) [reply src/maxtext/layers/moe.py:1537]: Potentially `sparse_matmul_route_and_compute` since we are inside of moe.py? We have sparse_matmul vs. dense_matmul strategies.

**Review by NuojCheng — APPROVED** [review]
Thank you for the efforts Shuwen!

File: src/maxtext/layers/moe.py [hunk 11]
```diff
@@ -55,6 +57,28 @@
...
+
+  expert_shard_id: int
+  local_sorted_indices: Optional[jax.Array]  # batch-sharded EP only
+  all_shards_group_sizes: Optional[jax.Array]  # batch-sharded EP only
+  reshaped_group_sizes: Optional[jax.Array]  # non-batch-sharded EP only
+
+
+@struct.dataclass
+class RouteOutput:
+  """Embed-independent routing state returned by route(). x and RouteMetadata are returned separately."""
```
Comment (gobbleturk) [inline src/maxtext/layers/moe.py]: I don't follow this comment at first (I need to read the code) - what is embed-indepedent routing state? Routing state except the actual activation?

File: src/maxtext/layers/moe.py [hunk 12]
```diff
@@ -1299,37 +1323,15 @@ def get_routed_moe_shardings(is_batch_sharded_by_expert):
...
-      batch_size, sequence_length, _ = x.shape
-      num_expert_parallelism = self.get_expert_parallelism_size()
-      if num_expert_parallelism > 1:
-        expert_shard_id = jax.lax.axis_index(self._expert_parallelism_name)
-      else:
-        expert_shard_id = 0
-      num_expert_parallelism = self.get_expert_parallelism_size()
-      global_group_sizes = None
+    def route(x, logits, pre_bias_logits, rngs):
+      """Route tokens by expert"""
```
Comment (gobbleturk) [inline src/maxtext/layers/moe.py]: I would modify comment to performs both across device and within device token routing/sorting

File: src/maxtext/layers/moe.py (hunk 2 above)
Comment (gobbleturk) [reply src/maxtext/layers/moe.py:1537]: sparse_matmul_route_and_compute sgtm

File: src/maxtext/layers/moe.py [hunk 13]
```diff
@@ -1455,42 +1463,69 @@ def get_active_sharding_axes(pspec_dim_axes, tensor_dim_index):
...
+        out_specs=(
+            self._logical_to_mesh_axes((batch_logical_axis, "activation_norm_length", "activation_embed")),
+            P(),  # Handle None or replicate the output
+            P(),  # Handle None or replicate the output
+        ),
+        check_vma=False,
+    )
+    def wrapper(x, logits, pre_bias_logits, w0, w1, wo, w0_bias, w1_bias, wo_bias, rngs):
+      batch_size, sequence_length, _ = x.shape
+      x, routing, route_metadata = route(x, logits, pre_bias_logits, rngs)
```
Comment (gobbleturk) [inline src/maxtext/layers/moe.py:1539]: yay this is great, so much complexity hidden in this function

File: src/maxtext/layers/moe.py [hunk 14]
```diff
@@ -1455,42 +1463,69 @@ def get_active_sharding_axes(pspec_dim_axes, tensor_dim_index):
...
+
+      gmm_fn = functools.partial(gmm, group_sizes=routing.group_sizes, expert_assignments=routing.selected_experts)
+      intermediate_layer = gmm_up(x, w0, w1, w0_bias, w1_bias, gmm_fn, weight_gather)
+
+      wo_gather_axes = []
+      if weight_gather:
+        # wo [Experts, Hidden, Out] -> Gather Exp(0) and Hidden(1)
+        wo_gather_axes.extend(get_active_sharding_axes(wo_pspec[0], 0))
+        wo_gather_axes.extend(get_active_sharding_axes(wo_pspec[1], 1))
+      wo_tile_size = (
```
Comment (gobbleturk) [inline src/maxtext/layers/moe.py]: for future PRs probably want another small function just to define this tile size, ideally the code inside of this sparse_matmul_route_and_compute function is as few lines as possible and easy to read - relying on calling smaller functions (go/small-functions)

**Review by gobbleturk — APPROVED** [review]

File: src/maxtext/layers/moe.py (hunk 14 above)
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py]: updated

File: src/maxtext/layers/moe.py (hunk 2 above)
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py:1537]: updated

File: src/maxtext/layers/moe.py (hunk 4 above)
Comment (RissyRan) [reply src/maxtext/layers/moe.py:1340]: I think we missed this check during migration. Could you help add them to [types.py](https://github.com/AI-Hypercomputer/maxtext/blob/d8763efa5bbc2d24b15fc5c856f885698f89e01e/src/maxtext/configs/types.py#L2886)?

https://github.com/AI-Hypercomputer/maxtext/blob/d8763efa5bbc2d24b15fc5c856f885698f89e01e/src/maxtext/configs/pyconfig_deprecated.py#L1213-L1215

**Review by RissyRan — APPROVED** [review]

File: src/maxtext/layers/moe.py (hunk 4 above)
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py:1340]: missed this, will add in next PR
